hbase-hadoop-compat/src/test/java/org/apache/hadoop/metrics2/impl/TestMutableRangeHistogram.java

```diff
@@ -0,0 +1,66 @@
+package org.apache.hadoop.metrics2.impl;
+
+import static org.junit.Assert.assertEquals;
+
+import org.apache.hadoop.hbase.HBaseClassTestRule;
+import org.apache.hadoop.hbase.testclassification.MetricsTests;
+import org.apache.hadoop.hbase.testclassification.SmallTests;
+import org.apache.hadoop.metrics2.AbstractMetric;
+import org.apache.hadoop.metrics2.MetricsRecordBuilder;
+import org.apache.hadoop.metrics2.lib.MutableRangeHistogram;
+import org.apache.hadoop.metrics2.lib.MutableSizeHistogram;
+import org.apache.hadoop.metrics2.lib.MutableTimeHistogram;
+import org.junit.ClassRule;
+import org.junit.Test;
+import org.junit.experimental.categories.Category;
+
+import java.util.Collection;
+
+@Category({ MetricsTests.class, SmallTests.class})
+public class TestMutableRangeHistogram {
+
+    @ClassRule
+    public static final HBaseClassTestRule CLASS_RULE =
+            HBaseClassTestRule.forClass(TestMutableRangeHistogram.class);
+
+    private static final String RECORD_NAME = "test";
+    private static final String SIZE_METRIC_NAME = "TestSize";
+    private static final String TIME_METRIC_NAME = "TestTime";
+
+    @Test
+    public void testMutableSizeHistogram() {
+        testMutableRangeHistogram(SIZE_METRIC_NAME, new MutableSizeHistogram(SIZE_METRIC_NAME, ""));
+    }
+
+    @Test
+    public void testMutableTimeHistogram() {
+        testMutableRangeHistogram(TIME_METRIC_NAME, new MutableTimeHistogram(TIME_METRIC_NAME, ""));
+    }
+
+    private static void testMutableRangeHistogram(String metricName, MutableRangeHistogram histogram) {
+        // fill up values
+        long[] ranges = histogram.getRanges();
+        for (long val : ranges) {
+            histogram.add(val - 1);
+        }
+        histogram.add(ranges[ranges.length - 1] + 1);
+        // put into metrics collector
+        MetricsCollectorImpl collector = new MetricsCollectorImpl();
+        MetricsRecordBuilder builder = collector.addRecord(RECORD_NAME);
+        histogram.snapshot(builder, true);
+        // get and print
+        Collection<MetricsRecordImpl> records = collector.getRecords();
+        assertEquals(records.size(), 1);
+        MetricsRecordImpl record = records.iterator().next();
+        assertEquals(record.name(), RECORD_NAME);
+        String prefix = metricName + "_" + histogram.getRangeType();
+        int count = 0;
+        for (AbstractMetric metric : record.metrics()) {
```

Comment:
here you could only validate the highest bins... depending on the precision. For `MutableTimeHistogram` with a precision of ~2000, you could probably validate these buckets: `10000, 30000, 60000, 120000, 300000, 600000`. The lower counts will eventually be incorrect due to the lack of precision.

The other option is to have several tests - each test could validate only some of the bins. For example, if you only fill in the lower bins, then the linear distribution of the internal bins of `FastLongHistogram` will be of smaller width, and the precision will be better for the lower buckets. You still need to `add` the values, take a `snapshot`, then `add` the values again, and take a `snapshot` gain, before doing the validation.

Reply:
I go it, thank you very much! I try to push again